src/java/com/puppetlabs/http/client/impl/JavaClient.java

```diff
@@ -299,11 +306,44 @@ else if (coercedOptions.getForceRedirects()) {
             redirectStrategy = new DefaultRedirectStrategy();
         }
         clientBuilder.setRedirectStrategy(redirectStrategy);
+
+        RequestConfig requestConfig = getRequestConfig(coercedOptions);
+
+        if (requestConfig != null) {
+            clientBuilder.setDefaultRequestConfig(requestConfig);
+        }
+
         CloseableHttpAsyncClient client = clientBuilder.build();
         client.start();
         return client;
     }
 
+    private static RequestConfig getRequestConfig
+            (CoercedClientOptions options) {
+        RequestConfig config = null;
+
+        int connectTimeoutMilliseconds = options.getConnectTimeoutMilliseconds();
+        int socketTimeoutMilliseconds = options.getSocketTimeoutMilliseconds();
+
+        if (connectTimeoutMilliseconds >= 0 || socketTimeoutMilliseconds >= 0) {
```

Comment:
This is a bit of an optimization.  I could instead just always construct a custom `RequestConfig` and only set the connect and socket timeouts on those if configured.  Apache has a memoized default `RequestConfig` that it uses when the client doesn't need to customize it.  The memoized version would likely load up a little faster, although I didn't do any formal performance characterization to validate that.